staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/capmanagers_test.go

```diff
@@ -0,0 +1,274 @@
+			expected: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 2, set("a", "b")),
+				entry("update-manager3", "v1", 3, set("c")),
+				entry("update-manager4", "v1", 4, set("d")),
+			},
+		}, {
+			name: "one version, one ancient changes",
+			input: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 2, set("a", "b")),
+				entry("update-manager3", "v1", 3, set("c")),
+				entry("update-manager4", "v1", 4, set("d")),
+				entry("update-manager5", "v1", 5, set("e")),
+			},
+			expected: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 3, set("a", "b", "c")),
+				entry("update-manager4", "v1", 4, set("d")),
+				entry("update-manager5", "v1", 5, set("e")),
+			},
+		}, {
+			name: "two versions, no ancient changes",
+			input: []metav1.ManagedFieldsEntry{
+				entry("update-manager1", "v1", 1, set("a")),
+				entry("update-manager2", "v2", 2, set("b")),
+				entry("update-manager3", "v1", 3, set("c")),
+				entry("update-manager4", "v1", 4, set("d")),
+				entry("update-manager5", "v1", 5, set("e")),
+			},
+			expected: []metav1.ManagedFieldsEntry{
+				entry("update-manager2", "v2", 2, set("b")),
+				entry("ancient-changes", "v1", 4, set("a", "c", "d")),
+				entry("update-manager5", "v1", 5, set("e")),
+			},
+		}, {
+			name: "three versions, one ancient changes",
+			input: []metav1.ManagedFieldsEntry{
+				entry("update-manager2", "v2", 2, set("b")),
+				entry("ancient-changes", "v1", 4, set("a", "c", "d")),
+				entry("update-manager5", "v1", 5, set("e")),
+				entry("update-manager6", "v3", 6, set("f")),
+				entry("update-manager7", "v2", 7, set("g")),
+			},
+			expected: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 5, set("a", "c", "d", "e")),
+				entry("update-manager6", "v3", 6, set("f")),
+				entry("ancient-changes", "v2", 7, set("b", "g")),
+			},
+		}, {
+			name: "three versions, two ancient changes",
+			input: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 5, set("a", "c", "d", "e")),
+				entry("update-manager6", "v3", 6, set("f")),
+				entry("ancient-changes", "v2", 7, set("b", "g")),
+				entry("update-manager8", "v3", 8, set("h")),
+			},
+			expected: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 5, set("a", "c", "d", "e")),
+				entry("ancient-changes", "v2", 7, set("b", "g")),
+				entry("ancient-changes", "v3", 8, set("f", "h")),
+			},
+		}, {
+			name: "four versions, two ancient changes",
+			input: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 5, set("a", "c", "d", "e")),
+				entry("update-manager6", "v3", 6, set("f")),
+				entry("ancient-changes", "v2", 7, set("b", "g")),
+				entry("update-manager8", "v4", 8, set("h")),
+			},
+			expected: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 5, set("a", "c", "d", "e")),
+				entry("update-manager6", "v3", 6, set("f")),
+				entry("ancient-changes", "v2", 7, set("b", "g")),
+				entry("update-manager8", "v4", 8, set("h")),
```

Comment:
In this test case, the number of entries after capping is >3, since there are 4 apiversions.